### ISSUE TITLE
Consider switching media queries to px breakpoints

### ISSUE THREAD

**Issue by CGamesPlay** [squidfunk/mkdocs-material#5530]
### Context

Material for MkDocs is a super widely used theme for lots of popular projects on the web. It's great to have this level of consistency amongst docs, and so I'm suggesting a small accessibility change that would have wide-reaching impact.

### Description

The breakpoints used in Material for MkDocs is presently [specified in em](https://github.com/squidfunk/mkdocs-material/blob/bf43e32e0b9e6b52a1016c3be13027b8ee4cd859/src/assets/stylesheets/_config.scss#L28). This decision was made 7 years ago, and I think it's worth revisiting now, specifically to switch to pixel units instead.

Switching to pixel units fixes an issue where the layout is broken on Safari for users with zoomed in/out pages. I ended up doing a bunch of research for this bug report, which I put into a [blog post](https://cgamesplay.com/post/2023/05/19/stop-using-em-in-media-queries/) where you can read more about the rationale. But the gist of it is: all major CSS frameworks and top internet sites are using px units for the media queries, and it avoids this browser bug, leading to increased accessibility for users.

### Related links

I've included a link to the relevant code to change and to a blog post with some additional explanation about the change and related browser bug.

### Use Cases

This browser bug affects Safari users (desktop and mobile, comprising 20% of global internet users) with a non-standard zoom level (presumably some small percentage of Safari users).

### Before submitting

- [X] I have read and followed the [change request guidelines](https://squidfunk.github.io/mkdocs-material/contributing/requesting-a-change/).
- [X] I have verified that [my idea is a change request and not a bug report](https://squidfunk.github.io/mkdocs-material/contributing/requesting-a-change/#its-not-a-bug-its-a-feature).
- [X] I have ensured that, to the best of my knowledge, [my idea will benefit the entire community](https://squidfunk.github.io/mkdocs-material/contributing/requesting-a-change/#benefit-for-the-community).
- [X] I have included relevant links to [the documentation](https://squidfunk.github.io/mkdocs-material), related [issues](https://github.com/squidfunk/mkdocs-material/issues), and [discussions](https://github.com/squidfunk/mkdocs-material/discussions) to underline the need for my idea.

**Comment by squidfunk** [comment]
Thanks for suggesting. From my knowledge, such a change would have very deep implications. We're using `rem` throughout our project to define all metrics based on text size. This means that when a user uses a larger font, the UI adapts accordingly. Using `rem` mandates using `em` in media queries – AFAIK there's no way around it. This means that if we decide do switch to `px` media queries, we would have to also drop `rem` based layouts, because with text zoom, `px` based media queries will be wrong. This also means that we can't do things like zoom the entire website:

https://github.com/squidfunk/mkdocs-material/blob/bf43e32e0b9e6b52a1016c3be13027b8ee4cd859/src/assets/stylesheets/main/components/_base.scss#L27-L48

Another option to solve this might be to use [media range syntax](https://caniuse.com/css-media-range-syntax), which also fixes the problem. Support is already good (not perfect, though). Might be worth exploring.

That being said, with my current knowledge, I _currently_ don't consider your proposal to be a good idea. However, if you would like to work on this and demonstrate that it is, I'm happy to discuss in a PR. If it turns out that we don't loose any of the current functionality, I'm happy to consider the switch, but as said, it can have very far reaching implications. I'd expect that this is tough and serious work. Would you be up to exploring possible directions?

**Comment by squidfunk** [comment]
In your article you write:

> Overall, the accessibility situation on the web is better than ever, but unfortunately we still live in a world with browser bugs. If you want your site to render the way you intended to users with low vision, you should keep this in mind and use em and rem units everywhere--except your media queries. Keep those in px.

IMHO, the following snippet shows why this is not a good idea – if you scale elements with font size, and don't do the same with media queries, your media queries will be off:

``` html
<html>
  <head>
    <style>
      body {
        margin: 0;
      }
      @media (min-width: 320px) {
        .foo {
          background: green;
          width: 320px;
        }
        .bar {
          background: red;
          width: 20rem;
        }
      }
    </style>
  </head>
  <body>
    <div class="foo">px</div>
    <div class="bar">em</div>
  </body>
</html>
```

The media query will _always_ breaks at 320px, but `bar` might be bigger, as it's based on font size.
<img width="1440" alt="Bildschirm­foto 2023-05-19 um 16 16 14" src="https://github.com/squidfunk/mkdocs-material/assets/932156/6f3b049a-fd60-439f-8e79-5f410812e76d">

<img width="1440" alt="Bildschirm­foto 2023-05-19 um 16 16 24" src="https://github.com/squidfunk/mkdocs-material/assets/932156/a392fa36-2813-4b25-b58a-ed902217dc54">

All in all, if we decide to switch to `px`, we need to consult accessibility experts. I'm not sure what the situation is and don't feel confident shipping a change of that magnitude, potentially locking out visually impaired people. If the consensus is that `px` are safe and font size zooming is not necessary at all, we can consider to do it.

**Comment by vedranmiletic** [comment]
> If the consensus is that `px` are safe and font size zooming is not necessary at all, we can consider to do it.

What do you mean by font size zooming not being necessary? For example, we tend to (ab)use Material for MkDocs as a slide deck presenter by zooming to 200-300% and scrolling the pages. It's sometimes more convenient than "normal" presentations exported to HTML via Marp or Quarto.

**Comment by squidfunk** [comment]
Not browser zooming, but font-size zooming, i.e., scale everything based on the default font size. See screenshots above.

**Comment by CGamesPlay** [comment]
Hey, I appreciate your thorough consideration. As a user who browses the web with an increased page zoom, I filed this issue because it would improve accessibility of many sites across the web, and I too definitely do not want to lock out visually impaired people 🙂

> Using rem mandates using em in media queries – AFAIK there's no way around it. This means that if we decide do switch to px media queries, we would have to also drop rem based layouts, because with text zoom, px based media queries will be wrong.

This isn't really how popular websites are doing things. For example, all of the Tailwind CSS [width classes](https://tailwindcss.com/docs/width) use rems to define width, but the media queries are nevertheless [specified in px](https://tailwindcss.com/docs/responsive-design). The [MDN sidebar](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/video) uses a media query specified in px but the uses rem for the spacing within the page.

Different browsers expose the font size setting in different ways. Chrome and Firefox both expose a "default font size" setting, but Safari only exposes a "minimum font size" setting, which has a different (worse) effect on page layouts. All browsers encourage users to use full page zoom rather than text size zooming, by making the setting equally easy or easier to access in all cases. For example, the menus for Chrome and Safari only have a full page zoom option (Firefox has a toggle-able option in the menu).

I did some experiments to see what the effect of the change is. I collected a series of screenshots showing Material for MkDocs' own documentation, with both types of zoom, before and after converting media queries to px-based. Here are the results:

- In all configurations, all Chrome and Firefox render the page the same. Safari is different, which is what this is all about.
- At the default zoom settings, there is no change to the appearance of the site (on any browser).
- Using page zoom, there is no change to the appearance of the site, except for Safari, which changed from a single-column view to the same two-column view seen in the other browsers. 
[Using em](https://github.com/squidfunk/mkdocs-material/assets/1216626/dc3f8595-8052-4f6d-b37c-7b5595b2c178) / [Using px](https://github.com/squidfunk/mkdocs-material/assets/1216626/b8587b28-e376-4279-ba4e-1f635cb50d3a)
- Using text only zoom, there is a layout change. [Using em](https://github.com/squidfunk/mkdocs-material/assets/1216626/a8750df9-68a0-4e2d-8274-3c18f25517b6) / [Using px](https://github.com/squidfunk/mkdocs-material/assets/1216626/e5e0ba29-f03f-41b3-9ad1-b5032b7872ba)
- Safari has no setting for default font size, only for minimum font size, which make this layout worse regardless of the media queries: [Using em](https://github.com/squidfunk/mkdocs-material/assets/1216626/bcc4ca55-9fc5-4895-88eb-3704f0078c87) / [Using px](https://github.com/squidfunk/mkdocs-material/assets/1216626/fb242649-6339-48af-acec-db4c3a63e8b8)

To briefly summarize where we're at so far, this change doesn't change anything except for fixing the browser bug and changing the way that text-only zoom renders the layouts. So I investigated this next. The question is more subjective: does this change reduce accessibility for users who are using text-only zoom? I tested different text-only zoom settings and compared the results.

- Maximized browser window (1785px)
  - Default font size: 16px ("Normal")
    - Both em and px render identically, using the 3-column layout ("default configuration").
  - Default font size: 20px ("Large")
    - Both em and px render identically, which is that the entire page is uniformly zoomed, indistinguishable from full page zoom.
  - Default font size: 24px ("Very Large")
    - Using em, the left sidebar collapses. The number of characters per line is about 30% more than the default configuration.
    - Using px, both sidebars remain. The number of characters per line is about 20% less than the default configuration.
- Half-screen browser window (879px)
  - In all font size options, both em and px render identically, with both sidebars collapsed.

Are there other configurations where there are more differences with em and px media queries? Yes, as demonstrated by your screenshot, it's always possible to design a scenario (combination of browser width and font size) where one of em or px is subjectively worse than the other.  However, I hope this series of experiments demonstrates that for likely configurations, this change does not harm the accessibility of the site.

---

In conclusion:
- Both px and em media queries render identically for the default configuration in all browsers.
- Only em media queries render incorrectly for users of Safari who use full-page zooming.
- Both px and em media queries have subjective differences in readability at specific browser widths, but both are always readable.
- I was not able to find any major website that used em media queries, even among websites that use rem-based layouts.

Please let me know if you have any other concerns. I believe this is a pretty compelling argument that we should switch to using px media queries.

**Comment by CGamesPlay** [comment]
Here's all of the screenshots I took for the tests in a ZIP file for comparison. [material tests.zip](https://github.com/squidfunk/mkdocs-material/files/11521853/material.tests.zip)

**Comment by squidfunk** [comment]
Thanks for your further input. Your tests look reasonable, but I'm on the fence whether we should mix `rem` and `px`. I don't think this is a good idea as they have different semantics, and mixing them might lead to weird layout bugs in certain configurations. Thus, I think we should either fully switch to `px` or keep `(r)em`. I'm not sure the Safari bug justifies making the switch, because it only impacts a very, very small amount of users (Safari users that use text zoom and are on a very specific resolution). Yes, the bug sucks, but there might also be other ways to mitigate it by inverting the logic of media queries, e.g. only using `max-` or `min-width`, but not mixing it or using it together.

> This isn't really how popular websites are doing things. 

I'm sorry, but I won't consider this being an argument. By that logic, you can also say that polluting the air is fine, because so many megacorps are doing it. Material for MkDocs took certain decisions that were or are harder to implement and maintain, but have upsides that other projects/websites cannot benefit from. For example, Material for MkDocs is fully built on the principle of progressive enhancement. I invite you to try another static site generator that is not and disable JavaScript. Some of them are barely usable. That is just not our philosophy.

That being said, we're currently not considering this change, because we are currently not regarding this as a good idea for the reasons stated and for the investment it would require to make the change. Unexpected breakage may occur, as many users have customized the theme. Also, the upside is rather small, and there might be easier ways to mitigate it without losing the benefits of `rem` based designs. If you, despite that, would like to work on this, I invite you to propose a PR. If you don't want to, please understand that our resources and time are also limited, and we always try to work on the parts of this project that benefit the most users. In the end, it's an economical decision.

**Comment by CGamesPlay** [comment]
It's disappointing that you feel that way. It's always nice to be able to help other people with things that I do to help myself, but I guess this change is too invasive for you. The commit that would perform this change is here, in case anyone wants to take this up. https://gist.github.com/CGamesPlay/77a4b18db9e48dcb4d57e9db76600255

**Comment by squidfunk** [comment]
> It's disappointing that you feel that way. It's always nice to be able to help other people with things that I do to help myself, but I guess this change is too invasive for you.

As said, if you or somebody else would like to take up the work, I'm happy to discuss over in a PR, so we have something concrete to talk about, but only switching media queries to `px` can't be merged, because now, all dimensions are still aligned with font size, but media queries are not. Thus, this is only half of what would need to be done.

**Comment by squidfunk** [comment]
Potential fix in 442bd1b07 using subpixel values for media queries when computing ems.

@CGamesPlay could you please check if this fixes your issues?

**Comment by CGamesPlay** [comment]
No, layout is still broken on Safari 16 unless initial page zoom is set to 100%.

**Comment by squidfunk** [comment]
Thanks for checking. Not sure why, I'm not able to reproduce it anymore. Starting at 100%, I zoomed three times and checked the 1220px breakpoint, for which the sidebar navigation was broken:

![Bildschirm­foto 2023-10-04 um 09 48 22](https://github.com/squidfunk/mkdocs-material/assets/932156/fcc56cab-fd16-4ef9-ba0b-fee9b52d56d1)

![Bildschirm­foto 2023-10-04 um 09 48 36](https://github.com/squidfunk/mkdocs-material/assets/932156/938c8230-aa96-4c56-98b5-0a2c9a33ec60)

![Bildschirm­foto 2023-10-04 um 09 48 45](https://github.com/squidfunk/mkdocs-material/assets/932156/310f4785-5fbf-48ed-847e-ec2d760bfd97)

Before the fix, the issue was observable, now I can't observe it anymore. I'm using Safari 16.1, macOS.

**Comment by CGamesPlay** [comment]
Just to confirm, the issue only happens if the page is zoomed on the initial page load. So the repro procedure would be to zoom the page, then refresh. At that point the media breakpoints will be set to `X * 16px * page zoom level` where X is the "em" value. This scaling only happens when the unit specified is "em", and is converted to px on the initial page load (so you have to zoom out twice as far to reach the breakpoint, or reload the page at 100% page zoom).

**Comment by squidfunk** [comment]
I'm not sure I understand, because I'm still not able to observe the problem. Could you please create a video recording, so that we're sure we're talking about the same thing?

**Comment by CGamesPlay** [comment]
[Here's a video](https://seafile.cluster.cgamesplay.com/f/6d0129e74ced4c55aba4/?dl=1) recording showing it happening on the current deployment of mkdocs-material. (This video link will expire in 90 days; sorry)

I appreciate you coming back to this issue.

**Comment by squidfunk** [comment]
Thanks for providing the video! I know what you're referring to now and I can confirm this behavior. No other browser shows it. What we fixed in the commit I linked above is broken layout on specific combinations of screen width and zoom level, that especially Safari was susceptible to, but I don't think this issue can be fixed with plain CSS, except for when we'd switch to `px` for all media queries. That being said, this issue does not lead to broken layouts, nor does it limit the zooming significantly. The main thing that's odd is that the drawer is hidden earlier, which is annoying but not severe, IMHO.

As we already discussed, this is not feasible at the current stage of the project – it would break too much, as we'd need to switch everything to `px`. We might consider going the all-`px` route in the future in a major release, but given how many sites Material for MkDocs is deployed, and how many people are using it, this needs to be done very carefully.

**Comment by CGamesPlay** [comment]
Understand your perspective and am not going to push on it any more. However, I suggest reevaluating the requirement that "px media queries mandate using all-px layouts". This is not a decision that Tailwind CSS, Bootstrap, or MUI make, nor is the decision that Google Search, YouTube, Facebook, MDN, CSS-Tricks, nor CodePen use on their production sites (they use px media queries and rem-based layouts).

**Comment by squidfunk** [comment]
It breaks font-zoom, as I showed in an earlier comment. If we drop the requirement to adhere to font-zoom, it can surely be done, but we need to assess this carefully.
